[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75260
* #75259
* #75258
* #75241
* #75202
* #75135
* __->__ #74882

Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

jerryzh168 committed:
[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict

Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented:
> when number of tensor inputs is 1,
we will share the output observer with input

I had a question about this.  If at a later time we want to improve this numerically, is the framework going to extend easily to support this?  For example, let's say there are some cases when we can just move the zero_point for scalar_add instead of recalculating the scale.  It's not p0 to do it now, but it would be good to have a path to do it later for edge casey things.

vkuzo commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -78,6 +82,31 @@ def _get_default_op_backend_config(op, dtype_configs):
     ]]
 
+
+def _binary_op_observation_type_getter(quantize_handler):
```
is this exposed to users?  Is there a way to not expose things like a quantize_handler input?

have we considered just modeling add_scalar and add_tensor separately in these configs?

jerryzh168 replied:
this is a custom function to overwrite the observation type, basically user can write a function to return the observation type based on the properties of root node. currently we use quantize_handler object, but this is just a convenience so that we don't need to spell out all relevant components here like root_node, modules etc., we can also expose root_node, modules, but I think that is not very scalable, since maybe there are other information that is relevant for user to decide the observation type.

jerryzh168 replied:
quantize_handler is just a place holder for the information that is related to the matched node here, we may rename it later

vkuzo replied:
it would be simpler to only allow the user to provide static configs, adding an extension point of a python function adds significant complexity. Can we allow different behaviors based on number of tensor arguments, and not expose the logic which calculates the number of tensor arguments?

jerryzh168 replied:
yeah sounds good

jerryzh168 replied:
will continue the discussion in the other comment

vkuzo commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -78,6 +82,31 @@ def _get_default_op_backend_config(op, dtype_configs):
+        return ObservationType.OUTPUT_SHARE_OBSERVER_WITH_INPUT
+
+_ADD_CONFIG = {
+    "pattern": operator.add,
```
this operator is overloaded, should we make the configs match the quantized ops (multiple configs, one per overload), instead of forking a single op config to handle the overloads?

jerryzh168 replied:
are you talking about `torch.add`? like https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L442 and https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L501?
I feel in general the overload info is not available to us, from our perspective we only see a floating point add operator in the graph, are you suggesting to infer the overload based on the dtype input of the operator?

jerryzh168 replied:
we have some design that takes types like (Tensor, Scalar) (maybe enum) before, but I'm not aware of an easy way to enable these things in the current codebase, I think this will require type inference, which we don't have currently. I'm open to change the api/implementation after we have the type inference ready in the future

vkuzo replied:
ok, how about just enabling different settings based on number of tensor arguments

jerryzh168 replied:
sure, we can do that, would this be to specific to the use case we have? do we plan to change it to specifying types after we have dtype inference?

vkuzo replied:
looks like the complexity is coming from here:
1. `torch.add` is a single operator in user code
2. `torch.add` calls different kernels (in fp32 or int8 or other dtypes) depending on # of tensor arguments

Do you want your config design to be unique by the operator type in user code (and have to support forking of each entry), or do you want your config design to be unique by kernel (and have to support one to many map of operator in user code to config entry)?  I think they are both correct, just a matter of which one you think fits better in your design.

jerryzh168 replied:
Yes, I think so. I feel it makes more sense to config the design to be unique by operator type/fusion pattern, since that is what we are matching against, otherwise the matching logic will be complex and will be subject to future extensions, e.g. if we have other variants that depend on other things, the matching logic need to change again to accommodate the changes.
This choice will simplify our implementation and make our matching code much more maintainable and understandable.

jerryzh168 replied:
using "_is_binary_op_with_binary_scalar_op_variant" and hided the internal implementation details, please take a look again

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35236032).

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 commented:
> > when number of tensor inputs is 1,
> > we will share the output observer with input
> 
> I had a question about this. If at a later time we want to improve this numerically, is the framework going to extend easily to support this? For example, let's say there are some cases when we can just move the zero_point for scalar_add instead of recalculating the scale. It's not p0 to do it now, but it would be good to have a path to do it later for edge casey things.

could you elaborate a bit on the use case? I think we are moving the zero_point right now for scalar add? https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qadd.cpp#L52

vkuzo commented:
> > > when number of tensor inputs is 1,
> > > we will share the output observer with input
> > 
> > 
> > I had a question about this. If at a later time we want to improve this numerically, is the framework going to extend easily to support this? For example, let's say there are some cases when we can just move the zero_point for scalar_add instead of recalculating the scale. It's not p0 to do it now, but it would be good to have a path to do it later for edge casey things.
> 
> could you elaborate a bit on the use case? I think we are moving the zero_point right now for scalar add? https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qadd.cpp#L52

if we are moving the zero point, why is the observer shared between input and output?  Would it be more correct to not have an observer at the output at all, since that formula assumes correct scale+zp of the input, and calculates scale+zp of the output based on scale+zp of the input?

jerryzh168 commented:
> > > > when number of tensor inputs is 1,
> > > > we will share the output observer with input
> > > 
> > > 
> > > I had a question about this. If at a later time we want to improve this numerically, is the framework going to extend easily to support this? For example, let's say there are some cases when we can just move the zero_point for scalar_add instead of recalculating the scale. It's not p0 to do it now, but it would be good to have a path to do it later for edge casey things.
> > 
> > 
> > could you elaborate a bit on the use case? I think we are moving the zero_point right now for scalar add? https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qadd.cpp#L52
> 
> if we are moving the zero point, why is the observer shared between input and output? Would it be more correct to not have an observer at the output at all, since that formula assumes correct scale+zp of the input, and calculates scale+zp of the output based on scale+zp of the input?

Yeah currently we do not model the numerics for add_scalar op correctly, we can't really model a change of zero_point  right now I think or it might require some more thinking. Currently sharing observer would help simplify the lowering for add_scalar, since it would have very similar pattern as the normal add.

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -78,6 +78,16 @@ def _get_default_op_backend_config(op, dtype_configs):
+_ADD_CONFIG = {
+    "pattern": operator.add,
+    "observation_type": ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT,
+    "_is_binary_op_with_binary_scalar_op_variant": True,
```
Thoughts on why there is an underscore?  It's ok to make this public.

jerryzh168 replied:
just feel this is temporay and we may have more general api after the dtype inference support

jerryzh168 replied:
this is changed to a map from num_tensor_args to the observation type

vkuzo commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -78,6 +78,16 @@ def _get_default_op_backend_config(op, dtype_configs):
 
+_ADD_CONFIG = {
+    "pattern": operator.add,
+    "observation_type": ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT,
```
looks like this is not always true, can we also put the other observation type in this config instead of hiding it?

vkuzo replied:
It could be something like

```
{
  "observation_type": ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT,
  "observation_type_if_binary_op_and_one_input_scalar": ObservationType.OUTPUT_SHARE_OBSERVER_WITH_INPUT,
}
```

or

```
{
  "observation_type": ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT_EXCEPT_ONE_INPUT_SCALAR_THEN_SHARE_OBSERVER_WITH_INPUT,
}
```

we should also make sure we have a way to change observation type of scalar add in the future if we need to

jerryzh168 replied:
maybe I can have a map from num_tensor_args to the observation type?

vkuzo commented on torch/ao/quantization/fx/backend_config/quantize_handler.py:
```diff
@@ -3,8 +3,29 @@
 
-def get_quantize_handler_cls(observation_type, dtype_configs):
+
+def _binary_op_observation_type_getter(quantize_handler):
```
I think this function should return the number of tensor arguments.  The logic of "if op is torch.add and num_tensor_args is 1, do this" should be visible from the config.

jerryzh168 replied:
yeah sounds good, I can change the config to a map

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/ns/fx/mappings.py:
```diff
@@ -450,7 +450,7 @@ def get_node_type_to_io_type_map() -> Dict[str, Set[NSNodeTargetType]]:
         # TODO(future PR): implement shadowing for binary ops and
         # uncomment below
-        # operator.add,
+        operator.add,
```
Hi @vkuzo is this OK? or should I skip the check in numeric suite test

vkuzo replied:
sg

andrewor14 replied:
Is the above TODO out of date then?

jerryzh168 replied:
there is a mul below still not enabled yet, I can move the comment

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35236032).

vkuzo commented on torch/ao/quantization/fx/backend_config/native.py:
```diff
@@ -78,6 +78,20 @@ def _get_default_op_backend_config(op, dtype_configs):
+    "pattern": operator.add,
+    "num_tensor_args_to_observation_type": {
+        # TODO: maybe change this to NO_OBSERVER
+        0: ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT,
```
nit: does the zero case actually appear, or does FX inline the calculation?

vkuzo replied:
if there are zero tensors, should we not insert observers then? Observers expect a tensor, violating this will lead to a runtime error.

jerryzh168 replied:
oh, yeah we do not insert observer actually, this is not really used right now since we have some extra checks in prepare: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L316, we can change this to NO_OBSERVER later after we have dtype inference implemented, maybe I can add more comments for TODO

andrewor14 commented on torch/ao/quantization/fx/quantization_patterns.py:
```diff
@@ -51,16 +51,22 @@ def __init__(
+                arg = self.root_node.args[arg_idx]
+                if isinstance(arg, Node) and (
+                        not all_node_args_have_no_tensors(
+                            arg, self.modules, cache_for_no_tensor_check)):
```
nit: This logic looks quite complicated. Maybe in a future PR we can rewrite `all_node_args_have_no_tensors` to instead return the number of tensor args instead, so we don't need to call this recursive function in a loop. I'm also not sure if we need the `cache` argument, or at least we should avoid exposing it to the caller.

jerryzh168 replied:
yeah both suggestions sounds good

jerryzh168 committed:
Update on "[quant][fx] Add support for BinaryOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinaryOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinaryOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinaryOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35236032).

jerryzh168 committed:
Update on "[quant][fx] Add support for BinaryOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35236032).

jerryzh168 committed:
Update on "[quant][fx] Add support for BinaryOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinaryOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx] Add support for BinarOpQuantizeHandler in backend_config_dict"


Summary:
This PR adds support for ops like add/mul in backend_config_dict, these ops have different
observation_type based on the number of tensor inputs, when number of tensor inputs is 1,
we will share the output observer with input, otherwise we'll have a new observer.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35236032](https://our.internmc.facebook.com/intern/diff/D35236032)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35236032).
